### PR TITLE
Fix shellcheck w/e in cluster/gce/gci/configure-helper.sh

### PR DESCRIPTION
**Pull request by joakimr-axis**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Fixes shellcheck failures in cluster/gce/gci/configure-helper.sh

**Which issue(s) this PR fixes**:
ref: #72956

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -182,13 +182,16 @@ function get-or-generate-uuid(){
...
+    local myuuid
+    myuuid=$(grep "${device}" "${ssdmap}" | cut -d ' ' -f 2)
   else
     # generate new uuid and add it to the map
-    local myuuid=$(uuidgen)
-    if [[ ! ${?} -eq 0 ]]; then
+    local myuuid
+    myuuid=$(uuidgen)
+    # shellcheck disable=SC2181
+    if [[ $? -ne 0 ]]; then
```
Comment (johscheuer) [inline cluster/gce/gci/configure-helper.sh:194]: Any reason you disabled the check here instead of using `if ! myuuid=$(uuidgen); then`?

Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:194]: No, your suggestion is what should be there. Thank you for this input. Update is coming.

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -182,13 +182,14 @@ function get-or-generate-uuid(){
...
     #create symlink based on saved uuid
-    local myuuid=$(grep ${device} ${ssdmap} | cut -d ' ' -f 2)
+    local myuuid
+    myuuid=$(grep "${device}" "${ssdmap}" | cut -d ' ' -f 2)
   else
     # generate new uuid and add it to the map
-    local myuuid=$(uuidgen)
-    if [[ ! ${?} -eq 0 ]]; then
+    local myuuid
+    if ! myuuid=$(uuidgen); then
```
Comment (phenixblue) [inline cluster/gce/gci/configure-helper.sh:192]: Does it make sense to simplify this slightly by moving the variable declaration outside the `if` statement, and then only worry about assignment within the if/else?

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -206,8 +207,10 @@ function get-or-generate-uuid(){
 #Formats the given device ($1) if needed and mounts it at given mount point
 # ($2).
 function safe-format-and-mount() {
-  local device="${1}"
-  local mountpoint="${2}"
+  local device
+  local mountpoint
+  device=$1
+  mountpoint=$2
```
Comment (phenixblue) [inline cluster/gce/gci/configure-helper.sh:213]: Probably want to carry the double quotes forward here

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -224,16 +227,19 @@ function safe-format-and-mount() {
 # Gets a devices UUID and bind mounts the device to mount location in
 # /mnt/disks/by-id/
 function unique-uuid-bind-mount(){
-  local mountpoint="${1}"
-  local actual_device="${2}"
+  local mountpoint
+  local actual_device
+  mountpoint=$1
+  actual_device=$2
```
Comment (phenixblue) [inline cluster/gce/gci/configure-helper.sh:233]: Same for double quotes

Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:233]: See previous comment.

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:213]: Quotes are superfluous for assignments.

```
$ a='foo bar'
$ b=$a
$ echo $b
foo bar
$
```

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:192]: I'm all for simplification, but I don't quite follow what you mean. Could you please explain with code to make it absolutely clear?

Comment (phenixblue) [reply cluster/gce/gci/configure-helper.sh:192]:
```
# each line of the ssdmap looks like "${device} persistent-uuid"
local myuuid
if grep -q "${device}" "${ssdmap}"; then
     #create symlink based on saved uuid
     myuuid=$(grep "${device}" "${ssdmap}" | cut -d ' ' -f 2)
else
     # generate new uuid and add it to the map
     if ! myuuid=$(uuidgen); then
         echo "Failed to generate valid UUID with uuidgen" >&2
         exit 2
     fi
....
```

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (phenixblue) [reply cluster/gce/gci/configure-helper.sh:213]: This is a practice defined by the style guide the project's coding conventions reference. Here's the specific section:

https://google.github.io/styleguide/shellguide.html#quoting

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:192]: Ah, yes, now I see. Yes, of course, very good suggestion. I will absolutely fix this.

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:212]: I read this:

```
# "prefer quoting words", not compulsory
readonly USE_INTEGER='true'

```

But if you want it, you shall have it. Update coming up.

File: cluster/gce/gci/configure-helper.sh [hunk 5]
```diff
@@ -242,16 +244,19 @@ function safe-format-and-mount() {
...
 
   # Trigger udev refresh so that newly formatted devices are propagated in by-uuid
   udevadm control --reload-rules
   udevadm trigger
   udevadm settle
 
   # grep the exact match of actual device, prevents substring matching
-  local myuuid=$(ls -l /dev/disk/by-uuid/ | grep "/${actual_device}$" | tr -s ' ' | cut -d ' ' -f 9)
+  local myuuid
+  myuuid=$(find -L /dev/disk/by-uuid -maxdepth 1 -samefile /dev/"${actual_device}" -printf "%P\n")
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:259]: can you explain why the change here? I'm not confident this is the same as before

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -335,7 +342,8 @@ function ensure-local-ssds() {
   local i=0
   for ssd in /dev/disk/by-id/google-local-ssd-*; do
     if [ -e "${ssd}" ]; then
-      local devicenum=`echo ${ssd} | sed -e 's/\/dev\/disk\/by-id\/google-local-ssd-\([0-9]*\)/\1/'`
+      local devicenum
+      devicenum=$(echo "${ssd}" | sed -e 's/\/dev\/disk\/by-id\/google-local-ssd-\([0-9]*\)/\1/')
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:346]: since we're touching this, let's make it more readable
```suggestion
      devicenum=$(echo "${ssd}" | sed -e 's@/dev/disk/by-id/google-local-ssd-\([0-9]*\)@\1@')
```

File: cluster/gce/gci/configure-helper.sh [hunk 7]
```diff
@@ -357,8 +365,9 @@ function ensure-local-ssds() {
     if [ -e "${ssd}" ]; then
       # This workaround to find if the NVMe device is a disk is required because
       # the existing Google images does not expose NVMe devices in /dev/disk/by-id
-      if [[ `udevadm info --query=property --name=${ssd} | grep DEVTYPE | sed "s/DEVTYPE=//"` == "disk" ]]; then
-        local devicenum=`echo ${ssd} | sed -e 's/\/dev\/nvme0n\([0-9]*\)/\1/'`
+      if [[ $(udevadm info --query=property --name="${ssd}" | grep DEVTYPE | sed "s/DEVTYPE=//") == "disk" ]]; then
+        local devicenum
+        devicenum=$(echo "${ssd}" | sed -e 's/\/dev\/nvme0n\([0-9]*\)/\1/')
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:370]: same here
```suggestion
        devicenum=$(echo "${ssd}" | sed -e 's@/dev/nvme0n\([0-9]*\)@\1@')
```

File: cluster/gce/gci/configure-helper.sh [hunk 8]
```diff
@@ -733,14 +743,15 @@ EOF
     use_cloud_config="true"
     if [[ -n "${NODE_TAGS:-}" ]]; then
       # split NODE_TAGS into an array by comma.
-      IFS=',' read -r -a node_tags <<< ${NODE_TAGS}
+      IFS=',' read -r -a node_tags <<< "${NODE_TAGS}"
     else
-      local -r node_tags="${NODE_INSTANCE_PREFIX}"
+      local -r node_tags
+      node_tags=${NODE_INSTANCE_PREFIX}
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:749]: style nit
```suggestion
      node_tags="${NODE_INSTANCE_PREFIX}"
```

File: cluster/gce/gci/configure-helper.sh [hunk 9]
```diff
@@ -1970,13 +1984,14 @@ function start-cluster-autoscaler {
...
-    create-kubeconfig "cluster-autoscaler" ${KUBE_CLUSTER_AUTOSCALER_TOKEN}
+    create-kubeconfig "cluster-autoscaler" "${KUBE_CLUSTER_AUTOSCALER_TOKEN}"
     prepare-log-file /var/log/cluster-autoscaler.log
 
     # Remove salt comments and replace variables with values
     local -r src_file="${KUBE_HOME}/kube-manifests/kubernetes/gci-trusty/cluster-autoscaler.manifest"
 
-    local params="${AUTOSCALER_MIG_CONFIG} ${CLOUD_CONFIG_OPT} ${AUTOSCALER_EXPANDER_CONFIG:---expander=price}"
+    local params
+    params="${AUTOSCALER_MIG_CONFIG}" "${CLOUD_CONFIG_OPT}" "${AUTOSCALER_EXPANDER_CONFIG:---expander=price}"
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:1994]: this doesn't look the same as before
```suggestion
    params="${AUTOSCALER_MIG_CONFIG} ${CLOUD_CONFIG_OPT} ${AUTOSCALER_EXPANDER_CONFIG:---expander=price}"
```

File: cluster/gce/gci/configure-helper.sh [hunk 10]
```diff
@@ -2918,7 +2937,7 @@ function main() {
     start-kube-controller-manager
     start-kube-scheduler
     wait-till-apiserver-ready
-    start-kube-addons
+    start-kube-addons "$@"
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:2940]: why do we need this?

File: cluster/gce/gci/configure-helper.sh [hunk 11]
```diff
@@ -453,7 +462,7 @@ function mount-master-pd {
   # locations.
 
   # Contains all the data stored in etcd.
-  mkdir -m 700 -p "${mount_point}/var/etcd"
+  install -d -m 700 "${mount_point}/var/etcd"
```
Comment (spiffxp) [inline cluster/gce/gci/configure-helper.sh:465]: why do we need this?  

if this in response to https://github.com/koalaman/shellcheck/wiki/SC2174 I'd rather be explicit about preserving the existing behavior instead of trying to "fix" it, unless you can point to bugs that are caused by this

**Review by spiffxp — CHANGES_REQUESTED** [review]
A few questions and suggestions

File: cluster/gce/gci/configure-helper.sh (hunk 5 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:259]: This patch is about taking care of the shellcheck warnings and errors of the current code. (Splitting the local part to a separate line goes without saying.)
The current ls | grep thing triggers [SC2010](https://github.com/koalaman/shellcheck/wiki/SC2010) and needs to be changed in order not to.
I figured instead of using these 4 different programs and 3 pipes, it would be more convenient to use 1 single well-known standard program (find) instead.

The current code does `ls -l` on `/dev/disk/by-uuid` which gives a result similar to

```
lrwxrwxrwx 1 root root 10 maj 16 08:18 0493e0d4-53c8-4f46-a339-f9e147a6e321 -> ../../dm-5
lrwxrwxrwx 1 root root 10 maj 16 08:18 1dcd0d80-2b96-4b1a-8890-881d20d59af6 -> ../../dm-6

```
[etc.]

and the greps `${actual_device} `to filter out the device we are searching the uuid for, pipes it to tr to replace all multiple occurrences of space with a single one and then pipes it to cut to use space as the delimiter and putput the 9th column, which is the uuid. For the `dm-5` above, this would be `0493e0d4-53c8-4f46-a339-f9e147a6e321`.

Now what the find command here does is to instead to search `/dev/disk/by-uuid` (with depth 1) and match the symlink with the `/dev/${actual_device}` that we are looking for (in this example above, match the symlink in `/dev/disk/by-uuid` with the `/dev/dm-5` which is `0493e0d4-53c8-4f46-a339-f9e147a6e321`, and then print the result with find's %P print format (file's name with the name of the starting-point under which it was found removed).

If you try both lines of code, you should see that they produce the same output. And, of course, the find solution passes shellcheck.

I would suggest that this is more robust than relying on the ls output.

File: cluster/gce/gci/configure-helper.sh (hunk 6 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:346]: Very good suggestion. I was not sure how popular deliminators other than / were here, so I didn't dare to change that. Update coming up!

File: cluster/gce/gci/configure-helper.sh (hunk 11 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:465]: You are correct that it is indeed SC2174 that I intended to mitigate here.
If you prefer the 2-liner suggested in https://github.com/koalaman/shellcheck/wiki/SC2174, I will of course update to that instead. (I chose install since it is a common tool and could be used in a oneliner fix, no other reason.)

File: cluster/gce/gci/configure-helper.sh (hunk 8 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:749]: Ah, yes, the superfluous assignment quotes.

File: cluster/gce/gci/configure-helper.sh (hunk 10 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:2940]: The way I understood it, the line where `start-kube-addons` does

`setup-node-termination-handler-manifest "$@"`

it is the script's inparameters that are used. But perhaps it is not, but rather eventual optional parameters sent to `start-kube-addons` and this line above should rather be

`start-kube-addons ''`

?

File: cluster/gce/gci/configure-helper.sh (hunk 9 above)
Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:1994]: Absolutely right, and should of course look as per your suggestion.

File: cluster/gce/gci/configure-helper.sh [hunk 12]
```diff
@@ -2423,7 +2440,7 @@ EOF
   fi
   if [[ "${ENABLE_NODE_TERMINATION_HANDLER:-}" == "true" ]]; then
       setup-addon-manifests "addons" "node-termination-handler"
-      setup-node-termination-handler-manifest
+      setup-node-termination-handler-manifest "$@"
```
Comment (cheftako) [inline cluster/gce/gci/configure-helper.sh:2443]: Where is $@ being populated from? Can we make this a little more clear? start-kube-addons does not seem to be taking an arguments.

Comment (joakimr-axis) [reply cluster/gce/gci/configure-helper.sh:2443]: Yes, it is unclear if it is ok to call setup-node-termination-handler-manifest without any arguments. I guess

`local -r nth_manifest="/etc/kubernetes/$1/$2/daemonset.yaml"`

would still work without $1 and $2, so no harm done if that is what was intended.
I think to keep the current behavior, this line should be

`setup-node-termination-handler-manifest '' ''`

And then an eventual refactorization is to be left as as separate task.

**Comment by cheftako** [comment]
Looks like there are still a few shell check errors.

**Comment by joakimr-axis** [comment]
There, now both shellcheck and all CI steps seem happy.

**Comment by spiffxp** [comment]
/approve
🤞

**Review by spiffxp — APPROVED** [review]
let's see what happens

**Comment by justaugustus** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-integration
(pod pending timeout)

**Comment by justaugustus** [comment]
/test pull-kubernetes-integration
(pod pending timeout)
